### PR TITLE
Added SubmissionForProcessing, submissions in queue in /submissions page and background job for automatic resubmission

### PR DESCRIPTION
**Pull request by PetarNeshkov**
Closes: 
https://github.com/SoftUni-Internal/exam-systems-issues/issues/794 
and
https://github.com/SoftUni-Internal/exam-systems-issues/issues/698

This pull request implements the following changes:

1. Usage of SubmissionsForProcessing table - Any new submission is saved in both the Submissions and SubmissionsForProcessing tables. The SubmissionsForProcessing table acts as a queue in the database. The following scenarios are possible for entries in SubmissionsForProcessing:
  - processing=false; processed=false - The submission was just added
  - processing=true; processed=false - The submission was sent to the message queue (status is "In Queue")
  - processing=false; processed=true - The execution result was received for the submission
- The submission is deleted from SubmissionsForProcessing after successfully processing the execution result

2. Pending submissions background job - A Hangfire job runs every 3 minutes to check the SubmissionsForProcessing table for pending submissions (processing=false and processed=false) that are older than 1 minute based on CreatedOn. This handles cases where there was an error publishing the submission to the queue (e.g. queue was down). The job publishes these submissions to RabbitMQ.

The background job is added via `GET` to https://admin.dev.alpha.judge.softuni.org/BackgroundJobs/AddRecurringEnqueuePendingSubmissionsForProcessingJob endpoint and can be monitored at the Hangfire dashboard https://admin.dev.alpha.judge.softuni.org/hangfire

3. FE changes

image

The following component should be visible to admins. Key points:
- Submissions awaiting execution - the count of submissions that are **procecessing (processing=true; processed=false)** or **pending (processing=false; processed=false)**
- All / In Queue/ Pending links for showing submissions in different states

### CODE REVIEW

File: Services/Infrastructure/OJS.Services.Infrastructure/Constants/CacheConstants.cs [hunk 1]
```diff
@@ -15,6 +15,7 @@ public static class CacheConstants
 
         public const string SubmissionsCountByMonthsForPastElevenMonthsKey = "Submissions_count_by_month_for_past_eleven_months";
         public const string SubmissionsCountForLastMonthKey = "Submissions_count_by_month_for_last_month";
+        public const string GetTotalSubmissionsCount = "GetTotalSubmissionsCount";
```
Comment (msmner) [inline Services/Infrastructure/OJS.Services.Infrastructure/Constants/CacheConstants.cs:18]: The name is not right for a constant.  There should be no verb like "get" there.

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs [hunk 2]
```diff
@@ -23,9 +23,14 @@ public SubmissionsForProcessingDataService(DbContext submissionsForProcessing)
...
 
+        public async Task<int> GetAllUnprocessedCount()
+            => await this.DbSet
+                .Where(sfp => !sfp.Processed && sfp.Processing)
+                .CountAsync();
+
         public IQueryable<SubmissionForProcessing> GetAllUnprocessed() =>
             this.DbSet
-                .Where(sfp => !sfp.Processed && !sfp.Processing);
+                .Where(sfp => !sfp.Processed && sfp.Processing);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs:33]: Why did you change this?

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs [hunk 3]
```diff
@@ -23,9 +23,14 @@ public SubmissionsForProcessingDataService(DbContext submissionsForProcessing)
                 .Where(s => s.SubmissionId == submissionId)
                 .FirstOrDefaultAsync();
 
+        public async Task<int> GetAllUnprocessedCount()
+            => await this.DbSet
+                .Where(sfp => !sfp.Processed && sfp.Processing)
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs:28]: You can reuse `GetAllUnprocessed`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

**Review by msmner — DISMISSED** [review]

File: Data/OJS.Data.Models/Submissions/SubmissionForProcessing.cs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 namespace OJS.Data.Models.Submissions
 {
+    using System;
     using SoftUni.Data.Infrastructure.Models;
 
     public class SubmissionForProcessing : Entity<int>
```
Comment (gogo4ds) [inline Data/OJS.Data.Models/Submissions/SubmissionForProcessing.cs:6]: Inherit `AuditInfoEntity<TId>`

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs [hunk 5]
```diff
@@ -86,10 +121,28 @@ public async Task ResetProcessingStatusById(int id)
             {
                 submissionForProcessing.Processing = false;
                 submissionForProcessing.Processed = false;
+                submissionForProcessing.CreatedOn = DateTime.Now.ToUniversalTime();
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs:124]: This will not be needed if entity is AuditInfoEntity

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs [hunk 6]
```diff
@@ -86,10 +121,28 @@ public async Task ResetProcessingStatusById(int id)
...
+
                 await this.SaveChanges();
             }
         }
 
+        public async Task MarkProcessed(int submissionId)
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs:130]: This seems the same as `EndProcessingBySubmission(int submissionId)`

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs [hunk 7]
```diff
@@ -50,20 +63,42 @@ public async Task AddOrUpdateBySubmissionIds(ICollection<int> submissionIds)
...
+        public async Task AddOrUpdateReprocessingBySubmission(int submissionId)
         {
             var submissionForProcessing = await this.GetBySubmission(submissionId);
 
             if (submissionForProcessing != null)
             {
-                submissionForProcessing.Processing = false;
+                submissionForProcessing.Processing = true;
                 submissionForProcessing.Processed = false;
+                submissionForProcessing.CreatedOn = DateTime.Now.ToUniversalTime();
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs:74]: Same here (AuditInfoEntity)

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs [hunk 8]
```diff
@@ -50,20 +63,42 @@ public async Task AddOrUpdateBySubmissionIds(ICollection<int> submissionIds)
...
 
-        public async Task AddOrUpdateBySubmission(int submissionId)
+        public async Task AddOrUpdateReprocessingBySubmission(int submissionId)
         {
             var submissionForProcessing = await this.GetBySubmission(submissionId);
 
             if (submissionForProcessing != null)
             {
-                submissionForProcessing.Processing = false;
+                submissionForProcessing.Processing = true;
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs:72]: By default the submission should be Processing = false. Only when it is queued should be set to Processing = true

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs [hunk 9]
```diff
@@ -129,11 +131,30 @@ public async Task<IActionResult> SaveExecutionResult([FromBody] SubmissionExecut
...
+    /// <summary>
+    /// Gets all unprocessed submissions.
+    /// </summary>
+    /// <param name="page">The current page number.</param>
+    /// <returns>A page with unprocessed submissions.</returns>
     [HttpGet]
-    [ProducesResponseType(typeof(IEnumerable<SubmissionForPublicSubmissionsServiceModel>), Status200OK)]
-    public async Task<IActionResult> Public()
+    [ProducesResponseType(typeof(PagedResultResponse<SubmissionForPublicSubmissionsResponseModel>), Status200OK)]
+    public async Task<IActionResult> GetProcessingSubmissions([FromQuery]int page)
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs:154]: This should be restricted only for admins/lecturers

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs [hunk 10]
```diff
@@ -142,8 +163,17 @@ public async Task<IActionResult> Public()
...
+        => await this.submissionCache
+            .GetTotalCount()
+            .ToOkResult();
+
+    /// <summary>
+    /// Gets the count of all unprocessed submissions.
+    /// </summary>
+    [HttpGet]
+    [ProducesResponseType(typeof(int), Status200OK)]
+    public async Task<IActionResult> UnprocessedTotalCount()
```
Comment (gogo4ds) [inline Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs:175]: Same restriction here

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs (hunk 6 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs]: Yes, removed it

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs (hunk 3 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs]: Done

File: Data/OJS.Data.Models/Submissions/SubmissionForProcessing.cs (hunk 4 above)
Comment (shsimeonova) [reply Data/OJS.Data.Models/Submissions/SubmissionForProcessing.cs]: Done

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs (hunk 5 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs]: Done

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs (hunk 7 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs]: Done

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs (hunk 8 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs]: Done

File: Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs (hunk 10 above)
Comment (shsimeonova) [reply Servers/UI/OJS.Servers.Ui/Controllers/Api/SubmissionsController.cs:179]: Done

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs [hunk 11]
```diff
@@ -22,16 +23,43 @@ public SubmissionsForProcessingDataService(DbContext submissionsForProcessing)
...
         await this.DbSet
             .Where(sfp => sfp.Processing && !sfp.Processed)
             .Select(sfp => sfp.Id)
             .ToListAsync();
 
+    public async Task<IEnumerable<TServiceModel>> GetAllProcessing<TServiceModel>()
+        => await this.DbSet
+            .Where(sfp => !sfp.Processed && sfp.Processing)
+            .ToListAsync()
+            .MapCollection<TServiceModel>();
```
Comment (gogo4ds) [inline Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:49]: MapCollection should be before the ToList?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs [hunk 12]
```diff
@@ -101,4 +108,38 @@ public new async Task Update(SubmissionForProcessing submissionForProcessing)
         base.Update(submissionForProcessing);
         await this.SaveChanges();
     }
+
+    public async Task MarkProcessing(int submissionId)
+    {
+        var submissionForProcessing = await this
+            .DbSet
+            .Where(sp => sp.SubmissionId == submissionId)
+            .FirstOrDefaultAsync();
```
[the remark this answers is not in the document]
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:117]: Fixed

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs (hunk 11 above)
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:49]: Fixed

File: Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs [hunk 13]
```diff
@@ -101,4 +108,38 @@ public new async Task Update(SubmissionForProcessing submissionForProcessing)
...
+        if (submissionForProcessing == null)
+        {
+            return;
+        }
+
+        submissionForProcessing.Processing = false;
+        submissionForProcessing.Processed = true;
+
+        await this.Update(submissionForProcessing);
+        await this.SaveChanges();
```
[the remark this answers is not in the document]
Comment (shsimeonova) [reply Services/Common/OJS.Services.Common.Data/Implementations/SubmissionsForProcessingCommonDataService.cs:140]: Fixed

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs (hunk 2 above)
Comment (gogo4ds) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsForProcessingDataService.cs]: It's good now

**Review by gogo4ds — APPROVED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs [hunk 14]
```diff
@@ -31,5 +48,31 @@ public async Task ResetAllProcessingSubmissions()
...
+            var submissions = this.submissionsData
+                .GetByIds(submissionsForProcessing
+                    .Select(sp => sp!.SubmissionId)
+                    .ToList())
+                .ToList();
+
+            submissions
+                .ForEachSequential((submission) =>
+                    this.submissionsCommonBusinessService
+                    .PublishSubmissionForProcessing(submission.Map<SubmissionServiceModel>()))
```
Comment (gogo4ds) [inline Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs:73]: Can you MapCollection directly to SubmissionServiceModel, instead of doing Map one by one from materialzed submissions?

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsDataService.cs [hunk 15]
```diff
@@ -23,20 +26,45 @@ public SubmissionsDataService(DbContext db)
...
                 orderBy: s => s.Id,
-                descending: true)
-            .Select(s => s.Id)
-            .FirstOrDefaultAsync();
+                descending: true);
+
+        if (user.IsLecturer)
+        {
+            submissionsQuery = submissionsQuery.Where(s => s.Participant!.UserId == user.Id);
+        }
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsDataService.cs:58]: Why filter submissions of the lecturer only?

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs (hunk 14 above)
Comment (shsimeonova) [reply Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs:73]: Done

File: Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs [hunk 16]
```diff
@@ -31,5 +48,33 @@ public async Task ResetAllProcessingSubmissions()
...
+            {
+                return;
+            }
+
+            var submissions = this.submissionsData
+                .GetByIds(submissionsForProcessing
+                    .Select(sp => sp!.SubmissionId)
+                    .ToList())
+                .MapCollection<SubmissionServiceModel>()
+                .ToList();
```
Comment (MTsatsarov) [inline Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs:69]: I think there is no need to call ToList() twice.

File: Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs [hunk 17]
```diff
@@ -441,47 +455,103 @@ public async Task ProcessExecutionResult(SubmissionExecutionResult submissionExe
...
         submission.ProcessingComment = null;
 
+        using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
         if (executionResult != null)
         {
             ProcessTestsExecutionResult(submission, executionResult);
+
+            await this.submissionsForProcessingData.MarkProcessed(submission.Id);
             this.submissionsData.Update(submission);
             await this.submissionsData.SaveChanges();
```
Comment (MTsatsarov) [inline Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs:465]: I think the MarkProcessed contains SaveChanges which can potentially lead to a case where the submissionForProcessing is marked as processed, but the submission result is not presenting in the database

**Comment by shsimeonova** [comment]
This pull request implements the following changes:

Usage of SubmissionsForProcessing table - Any new submission is saved in both the Submissions and SubmissionsForProcessing tables. The SubmissionsForProcessing table acts as a queue in the database. The following scenarios are possible for entries in SubmissionsForProcessing:
processing=false; processed=false - The submission was just added
processing=true; processed=false - The submission was sent to the message queue (status is "In Queue")
processing=false; processed=true - The execution result was received for the submission
The submission is deleted from SubmissionsForProcessing after successfully processing the execution result
Pending submissions background job - A Hangfire job runs every 3 minutes to check the SubmissionsForProcessing table for pending submissions (processing=false and processed=false) that are older than 1 minute based on CreatedOn. This handles cases where there was an error publishing the submission to the queue (e.g. queue was down). The job publishes these submissions to RabbitMQ.
3. FE changes
image
The following component should be visible to admins. Key points:
- Submissions awaiting execution - the count of submissions that are **procecessing (processing=true; processed=false)** or **pending (processing=false; processed=false)**
- All / In Queue/ Pending links for showing submissions in different states

File: Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs (hunk 16 above)
Comment (shsimeonova) [reply Services/Administration/OJS.Services.Administration.Business/Implementations/SubmissionsForProcessingBusinessService.cs:68]: Fixed

File: Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs (hunk 17 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs:465]: This code is in a transaction scope so even if UpdateResults throws an exception, nothing will be saved in the db no matter how many times we call the SaveChanges method.

**Review by MTsatsarov — APPROVED** [review]

File: Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs [hunk 18]
```diff
@@ -441,47 +455,103 @@ public async Task ProcessExecutionResult(SubmissionExecutionResult submissionExe
         submission.Processed = true;
         submission.ProcessingComment = null;
 
+        using var scope = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
         if (executionResult != null)
         {
             ProcessTestsExecutionResult(submission, executionResult);
+
+            await this.submissionsForProcessingData.MarkProcessed(submission.Id);
```
Comment (gogo4ds) [inline Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs:463]: The submission should be marked as processed, even when there is `submissionExecutionResult.Exception`

**Review by gogo4ds — CHANGES_REQUESTED** [review]

File: Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsDataService.cs (hunk 15 above)
Comment (gogo4ds) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsDataService.cs:58]: Ok, reply again when it's done

Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Data/Implementations/SubmissionsDataService.cs]: Done

File: Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs [hunk 19]
```diff
@@ -441,47 +455,103 @@ public async Task ProcessExecutionResult(SubmissionExecutionResult submissionExe
...
             submission.CompilerComment = errorMessage;
+
             this.submissionsData.Update(submission);
             await this.submissionsData.SaveChanges();
         }
+
+        scope.Complete();
+        scope.Dispose();
+
+        await this.submissionsForProcessingData.RemoveBySubmission(submission.Id);
```
[the remark this answers is not in the document]
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs]: Done

File: Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs (hunk 18 above)
Comment (shsimeonova) [reply Services/UI/OJS.Services.Ui.Business/Implementations/SubmissionsBusinessService.cs:463]: Done
